### ISSUE TITLE
New revisions (that are not approved yet) should work immediately for the user who created them

### ISSUE THREAD

**Issue by texttheater** [georgjaehnig/serchilo-drupal#73]
It's no fun to create/improve a shortcut if you can't use it immediately, plus you can't test it.

Sure, you could clone the shortcut to your own namespace or first create it there, test it and then change the namespace to whatever is appropriate, but that's quite a hassle.

**Comment by georgjaehnig** [comment]
I understand the problem and I'm thinking how to solve it.

First, the problem can be only solved for registered users who are not moderators. Because for an anonymous user, I don't see an easy way to make a revisions work *only* for her (maybe only playing around with cookies / sessions...).
So we are talking about registered users who fix site commands. If they do this regularly, I'm much on favor of making them moderators, so they also can approve commands.

If we really solve this in the software, we would need to deal with the mirror table [`serchilo_shortcut`](
https://github.com/georgjaehnig/serchilo-drupal/blob/master/modules/serchilo/serchilo.install#L46-L131). This database table mirrors all the shortcuts out of the complex, internal Drupal node and field storage into one table, for faster access. Currently, only approved shortcut revisions are in there.

An idea could be: On saving a non-approved revision by user `paul`, we put the shortcut into `serchilo_shortcut` as a user shortcut of `paul`. Although, this could cause problems if `paul` already has another shortcut with the same keyword and argument count.

So I'm quite hesitant to go the technical way, and more in favor trying it via moderation and user management policy, i.e. making users faster to moderators.

**Comment by texttheater** [comment]
I see your point, still think this feature would be very nice to have. I'm also specifically thinking of new users who're trying out Serchilo. Might be especially frustrating for them if their edits don’t work immediately. So this feature could be important for engaging new users.

**Comment by georgjaehnig** [comment]
I have got now the following implementation idea:

Unapproved shortcuts become available when calling a shortcut. When a unapproved shortcuts matches a query, the a Serchilo page is shown to user. The page informs that an unapproved shortcut matches the query and displays its target URL. The user has 2 options:

1. Use the unapproved shortcut: Get redirected to its target URL. At the same time, the revision ID of this shortcut revision gets saved in a cookie variable `accepted_revisions` in the clients browser. If the same accepted shortcut revision ID matches next time, the user gets redirected instantly.

2. Not use the unapproved shortcut: The query gets sent again, this time looking only for approved shortcuts. If the shortcut has an older, approved version, it should get matched. At the same time, the revision ID of this shortcut revision gets saved in a cookie variable `declined_revisions` in the clients browser. If the same declined shortcut revision ID matches next time, the query gets sent again, this time looking only for approved shortcuts.

So in short: Any user can activate any unapproved shortcut revision on a per-device basis.
 What do you think?

**Comment by texttheater** [comment]
The message and per-device thing sounds complicated, but seems to be needed to give anonymous users full freedom.

You could automatically add a revision to the cookie when it is created, this way anonymous users could create a new revision and use it instantly without warning.

Also, not sure if you ever need to bother registered users with the warning. If it's their own revision, I would run it for them without confirmation. If it's somebody else's, they can always clone it if they really can't wait for moderator approval.

I have a lingering feeling that there should be a simpler way to do all this. One thing I never really liked was conflating the user shortcut <-> site shortcut distinction with namespaces, because they're really orthogonal IMHO. IIRC in Serchilo 1.0 you could create language-specific user shortcuts. I miss that possibility. Now a special user shortcut <-> site shortcut distinction is kind of creeping in again through this security mechanism...

Maybe it's good idea to give this issue some more time and thought, as you say.

**Comment by georgjaehnig** [comment]
> You could automatically add a revision to the cookie when it is created, this way anonymous users could create a new revision and use it instantly without warning.

You're right, the cookie variable could be updated right on revision creation.

> IIRC in Serchilo 1.0 you could create language-specific user shortcuts. I miss that possibility.

Hm, I can't remember of this. Maybe you could have be able to add a `language = de` attribute without a warning to your user command, but I can't see how this would affect the behavior. In Serchilo 1.0, you could use your user commands from `your-username.serchilo.net`, and language-specific commands from `any-language-code.serchilo.net`, so I don't know how you could combine them. And how does a use case look now where you are missing them?

> I have a lingering feeling that there should be a simpler way to do all this. 

Yes, me too, all the time. :)

Another idea I had is to use different "releases", like Debian: Stable and Testing. Stable would be the default, and into Testing anyone could opt-in: registered users with a user setting, anonymous user with a URL variable.

Editing the shortcut repository would be completely unrestricted and changes would go immediately into Testing. Into Stable, only approved shortcuts would go, once a moderator did so.

The first problem that I see here: The Shortcut repository would soon look different than Stable, even for some days or weeks when controversial changes would cause a debate. So one would need to somehow show these 2 different versions on the site, which would need a lot of new code to write.

**Comment by georgjaehnig** [comment]
Maybe it helps to step back and try to remember what the "security system" wants to prevent. To me, there are 2 basic approaches:

1. Prevent existing behaviour to be changed by unapproved edits.
2. Prevent malicious URLs from being added.

The current demo implements 1: Anyone can add new shortcuts, even without approval, which even work immediately - as long as their keyword is new, and thus no other shortcuts are disturbed, i.e. the behaviour of all other shortcuts remains the same. But therefore, any later keyword and namespace is restricted.

Approach 2 would only address the danger that someone introduces a harmful URL, thus would make URL addings and changes needing an approval. The upside:All edits to keywords, namespaces would go without approval. The downside: Adding any new Shortcut with a new URL would need an approval.

**Comment by texttheater** [comment]
> Hm, I can't remember of this. Maybe you could have be able to add a language = de attribute without a warning to your user command, but I can't see how this would affect the behavior. In Serchilo 1.0, you could use your user commands from your-username.serchilo.net, and language-specific commands from any-language-code.serchilo.net, so I don't know how you could combine them.

By prefixing your command, e.g. `en:w` - IIRC this also worked for user commands. There was not too much use for this though because most language-specific commands should be site commands.

> And how does a use case look now where you are missing them?

Creating/editing language-specific commands. With Serchilo 1.0, I would mostly just edit the site commands an be done because there was no security. With Serchilo 2.0, I first have to clone/create them in my own namespace and then propose to add them to the appropriate namespace, plus possibly suggest a name change if there was a clash in my own namespace. That's what I find awkward. And Serchilo 3.0 is actually fixing this now that I think about it, so we're on the right way. :)

I'd say go ahead with the warnings proposal, at least currently I have no better idea.

**Comment by georgjaehnig** [comment]
I want to tacke this again and especially implement the idea already [described above](https://github.com/centic9/jgit-cookbook/issues/5#issuecomment-51084491). I started a new ticket for this in #116.

**Comment by georgjaehnig** [comment]
Obsolete within successor [www.trovu.net](http://www.trovu.net/).
